examples/kafka/src/test/java/org/apache/apex/examples/kafka/kafkaexactlyonceoutput/ApplicationTest.java

```diff
@@ -0,0 +1,149 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.examples.kafka.kafkaexactlyonceoutput;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FilenameFilter;
+import java.io.IOException;
+
+import javax.validation.ConstraintViolationException;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.hadoop.conf.Configuration;
+
+import com.datatorrent.api.LocalMode;
+
+import info.batey.kafka.unit.KafkaUnit;
+import info.batey.kafka.unit.KafkaUnitRule;
+
+import static org.junit.Assert.fail;
+
+/**
+ * Test the DAG declaration in local mode.
+ */
+public class ApplicationTest
+{
+  private static final String directory = "target/kafka_exactly";
+  private String tuplesUntilKill;
+
+  private static final int zkPort = 2181;
+  private static final int brokerPort = 9092;
+
+  private static final Logger logger = LoggerFactory.getLogger(ApplicationTest.class);
+
+  // broker port must match properties.xml
+  @Rule
+  public KafkaUnitRule kafkaUnitRule = new KafkaUnitRule(zkPort, brokerPort);
+
+  //remove '@After' to keep validation output file
+  @Before
+  @After
+  public void cleanup()
+  {
+    FileUtils.deleteQuietly(new File(directory));
+  }
+
+  @Test
+  public void testApplication() throws IOException, Exception
+  {
+    try {
+      createTopics();
+
+      // run app asynchronously; terminate after results are checked
+      Configuration conf = getConfig();
+      LocalMode lma = LocalMode.newInstance();
+      lma.prepareDAG(new Application(), conf);
+      ValidationToFile validationToFile = (ValidationToFile)lma.getDAG().getOperatorMeta("validationToFile").getOperator();
+      LocalMode.Controller lc = lma.getController();
+      lc.runAsync();
+
+      // get messages from Kafka topic and compare with input
+      int count = 1;
+      int maxSleepRounds = 300;
+      while (!validationToFile.validationDone) {
+        logger.info("Sleeping ....");
+        Thread.sleep(500);
+        if (count > maxSleepRounds) {
+          fail("validationDone flag did not get set to true in ValidationToFile operator");
+        }
+        count++;
+      }
+      lc.shutdown();
+      Thread.sleep(1000);
```

Comment:
so the validation application knows when the kafka app is done writing. the time it takes depends on the chosen number of tuples to write. It also needs a little bit of time writing to file before checking it. Therefore the 'Thread.sleep(1000)'.